### DISCUSSION TITLE
NativeStore and ID overflow

### DISCUSSION THREAD

**Discussion by kenwenzel** [eclipse-rdf4j/rdf4j#3331] [Q&A] [answered]
Hello,

NativeStore internally uses int-based IDs for references to values (literals, blank nodes, IRIs).
Is it possible that the IDs may be subject to overflows?

Thank you and best regards,
Ken

**Comment by abrokenjester** [comment, marked as answer]
Theoretically, yes, but given that integer max is over 2 billion, and the scalability of the native store in practice tops out about around 500-700 million triples, there is little danger of hitting that overflow. Keep in mind that even _very_ large datasets will (typically) contain a much smaller number of distinct values, as quite a lot of triples will reuse existing values (especially subjects and properties).

**Reply by kenwenzel** [reply]
Thank you for the answer, Jeen.
Does the NativeStore some kind of housekeeping/GC for removing of unused values or does the value store grow indefinitely?
I am asking because we have a scenario where multiple blank nodes are created and are later removed on a regular basis (material lots in an automated picking system).
Thank you and best regards
Ken

**Reply by hmottestad** [reply]
From what I remember there is no cleanup.

**Reply by abrokenjester** [reply]
Good point actually, and @hmottestad  is correct, there is no cleanup procedure for this. There is a long-open issue ticket for this in fact: https://github.com/eclipse/rdf4j/issues/550.

**Reply by kenwenzel** [reply]
Would it be an option to use [LevelDB](https://github.com/pcmind/leveldb) for storing the values or as possible alternative for the NativeStore?

**Reply by hmottestad** [reply]
If you have experience working with LevelDB and would like to try to improve the NativeStore then it would be an awesome contribution. Another alternative is to use the ExtensibleStore[1] as a base to implement your own store that uses LevelDB as the storage engine. You can take a look at these testes here if you want to see something dead simple to get you started: [ExtensibleStoreImplForTests.java](https://github.com/eclipse/rdf4j/blob/main/core/sail/extensible-store/src/test/java/org/eclipse/rdf4j/sail/extensiblestore/ExtensibleStoreImplForTests.java) and [ExtensibleStoreConnectionImplForTests.java](https://github.com/eclipse/rdf4j/blob/main/core/sail/extensible-store/src/test/java/org/eclipse/rdf4j/sail/extensiblestore/ExtensibleStoreConnectionImplForTests.java)   and [NaiveHashSetDataStructure.java](https://github.com/eclipse/rdf4j/blob/main/core/sail/extensible-store/src/test/java/org/eclipse/rdf4j/sail/extensiblestore/NaiveHashSetDataStructure.java)

1: https://github.com/eclipse/rdf4j/tree/main/core/sail/extensible-store/src/main/java/org/eclipse/rdf4j/sail/extensiblestore

**Reply by kenwenzel** [reply]
Thank you for the pointers. I will take a look. If I find some time then I try to create a first PoC.
How does ExtensibleStore handle transactions? Is there already some logic in place to handle isolation and rollbacks? This is actually what worries me the most when thinking about the LevelDB-based implementation.

**Reply by hmottestad** [reply]
Handles up to READ COMMITTED for you. Including rollback.

**Reply by kenwenzel** [reply]
I've added a PR with an LMDB-based SAIL for further discussion: #3413
